### PR TITLE
mono: add licensing details

### PR DESCRIPTION
**Pull request by 06kellyjac**
###### Motivation for this change

It'd be nice to have mono built and cached so I've tried to add more accurate licensing.

The [LICENSE](https://github.com/mono/mono/blob/main/LICENSE) file is quite the mess but the [FAQ](https://www.mono-project.com/docs/faq/licensing) asserts the key licenses are mit bsd3 and asl20

One other matter is they say "api-documentation" is CC BY 4.0 but IDK if the manpages count towards that

I think with the statements in the FAQ we can safely build and distribute mono at the very least until notified otherwise.
We can also post an issue to get this double checked by the mono maintainers if we want a second pair of eyes

Some other concerns are, do the licenses vary depending on mono versions? why does the arch release list it as `GPL, LGPL2.1, MPL`.
Debian has a very detailed copyright and license analysis so we could use some license details from this? https://metadata.ftp-master.debian.org/changelogs//main/m/mono/mono_6.8.0.105+dfsg-3.2_copyright

cc: @thoughtpolice @obadz @vrthra

---

I've gone with a `/* explanation */ license` format which is quite non-standard and `nixpkgs-fmt` doesn't like it (although `nixpkgs-fmt` doesn't like other parts of the file either). I can fix this to be as below if required:

```
# explanation
license
```

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by 06kellyjac** [comment]
I've bumped the mono maintainers again for comment. It might be worth looking into what debian/ubuntu/arch do for mono

**Comment by obadz** [comment]
Looks OK to me.

**Comment by drupol** [comment]
Hi all,

What's the status of this draft PR ? Is there anything left to move it forward?

I checked how [Gentoo is doing](https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-lang/mono/mono-6.12.0.122.ebuild#n13) and there's multiple licenses.

On Arch, multiple licenses as well: https://archlinux.org/packages/extra/x86_64/mono/

I haven't checked on Debian/Ubuntu.

If this PR is not relevant anymore, do you mind closing it?

Thanks!

**Comment by 06kellyjac** [comment]
Yeah I'm not sure why I drafted it. I guess to make sure mono maintainers had plenty of time to review it in case of a fly-by review + merge

I'm pretty sure these details still remain relevant according to the docs I listed in the comments but both gentoo and arch have some licenses that are missing.
I'm inclined to ignore arch as they don't even include MIT/ISC but that is clearly the main license

For gentoo the missing licenses are `LGPL-2.1 GPL-2 NPL-1.1 Ms-PL GPL-2-with-linking-exception IDPL`

Firstly for licenses they all exist in `licenses.nix` apart from the gpl2 with "linking-exception", [IDPL](https://gitweb.gentoo.org/repo/gentoo.git/plain/licenses/IDPL) and [NPL](https://gitweb.gentoo.org/repo/gentoo.git/plain/licenses/NPL-1.1)

It looks like any reference to NPL is actually MPL
https://sourcegraph.com/github.com/mono/mono/-/blob/LICENSE?L399

There is general mention of GPL and explicit mention of GPLv2 once
https://sourcegraph.com/github.com/mono/mono/-/blob/LICENSE?L31

There is an LGPL mention here but it's v3:
https://sourcegraph.com/github.com/mono/mono/-/blob/mcs/tools/sqlmetal/src/DbLinq/Test/Providers/THIRD-PARTY-NOTICES.TXT?L18

There are a handful of general mentions of LGPL as being the license for some original code but they say they port is under MIT (not sure if that flys with GPL certainly 3 but :shrug:)
https://sourcegraph.com/github.com/mono/mono/-/blob/mcs/class/System.Data/Test/THIRD-PARTY-NOTICES.TXT?L35-36 (and others)

Looks like Fedora goes with just MIT https://support.atlassian.com/jira-software-cloud/docs/manage-epics-in-team-managed-projects/

---

So TL;DR:
I think what this PR proposes is pretty accurate and I can undraft it after some changes

Since GPLv2 is explicitly mentioned at one point in the LICENSE file I can add that one in too.
Not too sure about adding LGPL.
NPL seems covered by MPL 1.1 which is explicitly mentioned so I can add that.
And there were no matches for IDPL or "Initial Developer" outside of references in the MPL part
MSPL/Microsoft Public License is mentioned so I should add that too.

None of these licenses would mark mono unfree so wouldn't affect how the package is consumed outside of people who are explicit with the licenses they accept

**Comment by 06kellyjac** [comment]
I ran fossa against the project and it didn't really like that. I guess I was under the false impression that fossa would be able to handle things pretty well. It does seem fairly flexible in being able to correct false positives and (usually) see why some result appeared but the results weren't great. I guess it works better on projects that have dependency files e.g. Cargo.lock, package-lock.json

It either saw EnglishWords.txt mentions coveralls or some of the extenrnal submodules have coveralls badges in their readme & ran via travis so fossa pulled in GPLv3

image

I'm not sure what about "coveralls.io" made it think GPLv3 considering the backend seems to be closed source & their client tools seem to be MIT lol

**Comment by 06kellyjac** [comment]
updated

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
Thanks ! Glad to have fixed this long overdue issue !
